lib/spack/spack/analyzers/analyzerbase.py

```diff
@@ -0,0 +1,105 @@
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""An analyzer base provides basic functions to run the analysis, save results,
+and (optionally) interact with a Spack Monitor
+"""
+
+import spack.monitor
+import spack.hooks
+import llnl.util.tty as tty
+import spack.util.path
+import spack.config
+
+import os
+
+
+def get_analyzer_dir(spec, analyzer_dir=None):
+    """Given a spec, return the directory to save analyzer results,
+    creating it if it does not exist. We also check that the spec has an
+    associated package. An analyzer cannot be run if the spec isn't
+    associated with a package. If the user provides a custom analyzer_dir,
+    we use it over checking the config and the default at ~/.spack/analyzers
+    """
+    # An analyzer cannot be run if the spec isn't associated with a package
+    if not hasattr(spec, "package") or not spec.package:
+        tty.die("A spec can only be analyzed with an associated package.")
+
+    # The top level directory is in the user home, or a custom location
+    if not analyzer_dir:
+        analyzer_dir = spack.util.path.canonicalize_path(
+            spack.config.get('config:analyzers_dir', '~/.spack/analyzers'))
+
+    # We follow the same convention as the spec install (this could be better)
+    package_prefix = os.sep.join(spec.package.prefix.split('/')[-3:])
+    meta_dir = os.path.join(analyzer_dir, package_prefix)
+    return meta_dir
+
+
+class Analyzerbase(object):
+
+    def __init__(self, spec, dirname=None):
+        """An Analyzer is intended to run on one spec install, so the spec
+        with its associated package is required on init. The child analyzer
+        class should define an init function that super's the init here, and
+        also check that the analyzer has all dependencies that it
+        needs. If an analyzer subclass does not have dependencies, it does not
+        need to define an init. An Analyzer should not be allowed to proceed
+        if one or more dependencies are missing. The dirname, if defined,
+        is an optional directory name to save to (instead of the default meta
+        spack directory).
+        """
+        self.spec = spec
+        self.dirname = dirname
+        self.meta_dir = os.path.dirname(spec.package.install_log_path)
+
+        for required in ["name", "outfile", "description"]:
+            if not hasattr(self, required):
+                tty.die("Please add a %s attribute on the analyzer." % required)
+
+    def run(self):
+        """Given a spec with an installed package, run the analyzer on it.
+        """
+        raise NotImplementedError
+
+    @property
+    def output_dir(self):
+        """The full path to the output directory, including the nested analyzer
+        directory structure. This will create the full path if it doesn't exist,
+        so you should be sure that you have a result before using the property.
```

Comment:
the auto-creation thing is a little weird -- I'd expect this to return the path regardless of whether it exists, and save_result to create the directory.  I think it'll be easy to create superfluous directories with the functionality merged like this. Is there a reason you want the property to create the directory?

Reply:
You can probably tell from my justification in the docstring that I thought about it, and did it this way (but wasn't very happy with the prospect of having directories created) and then added "you should have a result before using this property." I think you are probably right we should have the save_result function do the creation, I'll fix this up.

Reply:
This should also be fixed. The save_result (or wherever the analyzer needs to write files) is in charge of creation.